src/java.base/share/classes/jdk/internal/classfile/impl/DirectCodeBuilder.java

```diff
@@ -305,6 +308,54 @@ public void writeBody(BufWriter b) {
         }
 
         content = new UnboundAttribute.AdHocAttribute<>(Attributes.CODE) {
+
+            private void writeCounters(boolean codeMatch, BufWriterImpl buf) {
+                if (codeMatch) {
+                    buf.writeU2(original.maxStack());
+                    buf.writeU2(original.maxLocals());
+                } else {
+                    StackCounter cntr = StackCounter.of(DirectCodeBuilder.this, buf);
+                    buf.writeU2(cntr.maxStack());
+                    buf.writeU2(cntr.maxLocals());
+                }
+            }
+
+            private void generateStackMaps(BufWriterImpl buf) throws IllegalArgumentException {
+                //new instance of generator immediately calculates maxStack, maxLocals, all frames,
+                // patches dead bytecode blocks and removes them from exception table
+                StackMapGenerator gen = StackMapGenerator.of(DirectCodeBuilder.this, buf);
+                attributes.withAttribute(gen.stackMapTableAttribute());
+                buf.writeU2(gen.maxStack());
+                buf.writeU2(gen.maxLocals());
+            }
+
+            private void tryGenerateStackMaps(boolean codeMatch, BufWriterImpl buf) {
+                if (buf.getMajorVersion() >= Classfile.JAVA_6_VERSION) {
+                    try {
+                        generateStackMaps(buf);
+                    } catch (IllegalArgumentException e) {
+                        //failover following JVMS-4.10
+                        if (buf.getMajorVersion() == Classfile.JAVA_6_VERSION) {
+                            writeCounters(codeMatch, buf);
+                        } else {
+                            throw e;
+                        }
+                    }
+                } else {
+                    writeCounters(codeMatch, buf);
+                }
+            }
+
+            private boolean writeOriginalAttribute(BufWriterImpl buf) {
```

Comment:
As a result (see previous comment), we will run into a redundant stack map generation when the original stack map table is trivial (absent). Maybe we want to consider user-configured selective stack map regeneration to replace this faulty logiic.

Reply:
I've fixed the logic in the commit below.